File: cathode/src/main/java/net/simonvt/cathode/ui/dashboard/DashboardShowsWatchlistAdapter.java
/*
 * Copyright (C) 2016 Simon Vig Therkildsen
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *    http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package net.simonvt.cathode.ui.dashboard;

import android.content.Context;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.TextView;
import androidx.annotation.NonNull;
import androidx.recyclerview.widget.RecyclerView;
import butterknife.BindView;
import butterknife.ButterKnife;
import java.util.ArrayList;
import java.util.List;
import net.simonvt.cathode.R;
import net.simonvt.cathode.common.ui.adapter.BaseAdapter;
import net.simonvt.cathode.common.widget.RemoteImageView;
import net.simonvt.cathode.entity.Episode;
import net.simonvt.cathode.entity.Show;
import net.simonvt.cathode.images.ImageType;
import net.simonvt.cathode.images.ImageUri;
import net.simonvt.cathode.provider.util.DataHelper;

public class DashboardShowsWatchlistAdapter extends BaseAdapter<Object, RecyclerView.ViewHolder> {

  private static final int TYPE_SHOW = 0;
  private static final int TYPE_EPISODE = 1;

  private Context context;

  private DashboardFragment.OverviewCallback callback;

  private List<Show> shows;
  private List<Episode> episodes;

  public DashboardShowsWatchlistAdapter(Context context,
      DashboardFragment.OverviewCallback callback) {
    super(context);
    this.context = context;
    this.callback = callback;
  }

  @Override protected boolean areItemsTheSame(@NonNull Object oldItem, @NonNull Object newItem) {
    if (oldItem instanceof Show && newItem instanceof Show) {
      return ((Show) oldItem).getId() == ((Show) newItem).getId();
    } else if (oldItem instanceof Episode && newItem instanceof Episode) {
      return ((Episode) oldItem).getId() == ((Episode) newItem).getId();
    }

    return false;
  }

  public void changeShowList(List<Show> shows) {
    this.shows = shows;
    updateItems();
  }

  public void changeEpisodeList(List<Episode> episodes) {
    this.episodes = episodes;
    updateItems();
  }

  private void updateItems() {
    List<Object> items = new ArrayList<>();

    if (shows != null) {
      items.addAll(shows);
    }

    if (episodes != null) {
      items.addAll(episodes);
    }

    setList(items);
  }

  @Override public int getItemViewType(int position) {
    if (getList().get(position) instanceof Show) {
      return TYPE_SHOW;
    }

    return TYPE_EPISODE;
  }

  @Override public long getItemId(int position) {
    // TODO: Better way=
    if (getItemViewType(position) == TYPE_SHOW) {
      return ((Show) getList().get(position)).getTvdbId();
    }

    return ((Episode) getList().get(position)).getTvdbId();
  }

  @Override public RecyclerView.ViewHolder onCreateViewHolder(ViewGroup parent, int viewType) {
    if (viewType == TYPE_SHOW) {
      final View view = LayoutInflater.from(parent.getContext())
          .inflate(R.layout.list_row_dashboard_show, parent, false);
      final ShowViewHolder holder = new ShowViewHolder(view);

      view.setOnClickListener(new View.OnClickListener() {
        @Override public void onClick(View v) {
          final int position = holder.getAdapterPosition();
          if (position != RecyclerView.NO_POSITION) {
            Show show = (Show) getList().get(position);
            callback.onDisplayShow(show.getId(), show.getTitle(), show.getOverview());
          }
        }
      });

      return holder;
    } else {
      final View view = LayoutInflater.from(parent.getContext())
          .inflate(R.layout.list_row_dashboard_episode, parent, false);
      final EpisodeViewHolder holder = new EpisodeViewHolder(view);

      view.setOnClickListener(new View.OnClickListener() {
        @Override public void onClick(View v) {
          final int position = holder.getAdapterPosition();
          if (position != RecyclerView.NO_POSITION) {
            Episode episode = (Episode) getList().get(position);
            callback.onDisplayEpisode(episode.getId(), episode.getShowTitle());
          }
        }
      });

      return holder;
    }
  }

  @Override public void onBindViewHolder(RecyclerView.ViewHolder holder, int position) {
    if (holder.getItemViewType() == TYPE_SHOW) {
      ShowViewHolder showHolder = (ShowViewHolder) holder;
      Show show = (Show) getList().get(position);

      final String poster = ImageUri.create(ImageUri.ITEM_SHOW, ImageType.POSTER, show.getId());
      showHolder.poster.setImage(poster);
      showHolder.title.setText(show.getTitle());
    } else {
      EpisodeViewHolder episodeHolder = (EpisodeViewHolder) holder;
      Episode episode = (Episode) getList().get(position);

      final String screenshotUri =
          ImageUri.create(ImageUri.ITEM_EPISODE, ImageType.STILL, episode.getId());
      episodeHolder.screenshot.setImage(screenshotUri);

      final String title =
          DataHelper.getEpisodeTitle(context, episode.getTitle(), episode.getSeason(),
              episode.getEpisode(), episode.getWatched(), true);
      episodeHolder.title.setText(title);
    }
  }

  static class ShowViewHolder extends RecyclerView.ViewHolder {

    @BindView(R.id.poster) RemoteImageView poster;
    @BindView(R.id.title) TextView title;

    ShowViewHolder(View itemView) {
      super(itemView);
      ButterKnife.bind(this, itemView);
    }
  }

  static class EpisodeViewHolder extends RecyclerView.ViewHolder {

    @BindView(R.id.screenshot) RemoteImageView screenshot;
    @BindView(R.id.title) TextView title;

    EpisodeViewHolder(View itemView) {
      super(itemView);
      ButterKnife.bind(this, itemView);
    }
  }
}

File: cathode/src/main/java/net/simonvt/cathode/ui/dashboard/DashboardShowsWatchlistAdapter.kt
/*
 * Copyright (C) 2016 Simon Vig Therkildsen
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *    http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package net.simonvt.cathode.ui.dashboard

import android.content.Context
import android.view.LayoutInflater
import android.view.ViewGroup
import androidx.recyclerview.widget.RecyclerView
import androidx.recyclerview.widget.RecyclerView.ViewHolder
import net.simonvt.cathode.common.ui.adapter.BaseAdapter
import net.simonvt.cathode.databinding.ListRowDashboardEpisodeBinding
import net.simonvt.cathode.databinding.ListRowDashboardShowBinding
import net.simonvt.cathode.entity.Episode
import net.simonvt.cathode.entity.Show
import net.simonvt.cathode.images.ImageType.POSTER
import net.simonvt.cathode.images.ImageType.STILL
import net.simonvt.cathode.images.ImageUri
import net.simonvt.cathode.provider.util.DataHelper
import net.simonvt.cathode.ui.dashboard.DashboardFragment.OverviewCallback
import java.util.ArrayList

class DashboardShowsWatchlistAdapter(
  context: Context,
  private val callback: OverviewCallback
) : BaseAdapter<Any, ViewHolder>(context) {

  private var shows: List<Show>? = null
  private var episodes: List<Episode>? = null

  override fun areItemsTheSame(oldItem: Any, newItem: Any): Boolean {
    if (oldItem is Show && newItem is Show) {
      return oldItem.id == newItem.id
    } else if (oldItem is Episode && newItem is Episode) {
      return oldItem.id == newItem.id
    }
    return false
  }

  fun changeShowList(shows: List<Show>?) {
    this.shows = shows
    updateItems()
  }

  fun changeEpisodeList(episodes: List<Episode>?) {
    this.episodes = episodes
    updateItems()
  }

  private fun updateItems() {
    val items: MutableList<Any> = ArrayList()
    if (shows != null) {
      items.addAll(shows!!)
    }
    if (episodes != null) {
      items.addAll(episodes!!)
    }

    list = items
  }

  override fun getItemViewType(position: Int): Int {
    return if (list[position] is Show) {
      TYPE_SHOW
    } else TYPE_EPISODE
  }

  override fun getItemId(position: Int): Long {
    // TODO: Better way?
    return if (getItemViewType(position) == TYPE_SHOW) {
      (list[position] as Show).tvdbId.toLong()
    } else {
      (list[position] as Episode).tvdbId.toLong()
    }
  }

  override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): ViewHolder {
    if (viewType == TYPE_SHOW) {
      val binding =
        ListRowDashboardShowBinding.inflate(LayoutInflater.from(parent.context), parent, false)
      val holder = ShowViewHolder(binding)
      binding.root.setOnClickListener {
        val position = holder.adapterPosition
        if (position != RecyclerView.NO_POSITION) {
          val show = list[position] as Show
          callback.onDisplayShow(show.id, show.title, show.overview)
        }
      }
      return holder
    } else {
      val binding =
        ListRowDashboardEpisodeBinding.inflate(LayoutInflater.from(parent.context), parent, false)
      val holder = EpisodeViewHolder(binding)
      binding.root.setOnClickListener {
        val position = holder.adapterPosition
        if (position != RecyclerView.NO_POSITION) {
          val episode = list[position] as Episode
          callback.onDisplayEpisode(episode.id, episode.showTitle)
        }
      }
      return holder
    }
  }

  override fun onBindViewHolder(holder: ViewHolder, position: Int) {
    if (holder.itemViewType == TYPE_SHOW) {
      val showHolder = holder as ShowViewHolder
      val (id, title) = list[position] as Show
      val poster = ImageUri.create(ImageUri.ITEM_SHOW, POSTER, id)
      showHolder.poster.setImage(poster)
      showHolder.title.text = title
    } else {
      val episodeHolder =
        holder as EpisodeViewHolder
      val episode =
        list[position] as Episode
      val screenshotUri = ImageUri.create(ImageUri.ITEM_EPISODE, STILL, episode.id)
      episodeHolder.screenshot.setImage(screenshotUri)
      val title = DataHelper.getEpisodeTitle(
        context,
        episode.title,
        episode.season,
        episode.episode,
        episode.watched,
        true
      )
      episodeHolder.title.text = title
    }
  }

  class ShowViewHolder(binding: ListRowDashboardShowBinding) : ViewHolder(binding.root) {
    var poster = binding.poster
    var title = binding.title
  }

  class EpisodeViewHolder(binding: ListRowDashboardEpisodeBinding) : ViewHolder(binding.root) {
    val screenshot = binding.screenshot
    val title = binding.title
  }

  companion object {
    private const val TYPE_SHOW = 0
    private const val TYPE_EPISODE = 1
  }
}
